File: java/jsonfold-core/src/main/java/dev/jsonfold/format/Line.java
package dev.jsonfold.format;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Represents a single pretty-printed JSON line.
 *
 * <p>
 * Lines are the basic unit processed by the folding engine.
 * A line may represent:
 *
 * <ul>
 * <li>a scalar value or property</li>
 * <li>an opening container ({ or [)</li>
 * <li>a closing container (} or ])</li>
 * <li>a previously folded container</li>
 * </ul>
 *
 * <p>
 * Additional metadata tracks packing and folding eligibility.
 */
final class Line {

    /** Leading indentation width in spaces. */
    final int indent;

    /** For folded lines - the line kind */
    Kind kind = Kind.NONE ;

    List<String> parts = new ArrayList<>() ;
    int partsLength = 0 ;

    /** Line text without leading indentation or trailing newline. */
//    String text;

    /** Logical item count represented by this line. */
    int items ;

    /** Number of scalar leaf values represented by this line. */
    int leafs ;

    /**
     * Maximum folded child nesting represented by this line.
     *
     * <p>
     * A value of {@code -1} means this is a scalar/body line.
     */
    int childNesting = -1;

    /** Opening container kind, or {@link Kind#NONE}. */
    Kind opener = Kind.NONE;

    /** Closing container kind, or {@link Kind#NONE}. */
    Kind closer = Kind.NONE;

    /** Whether this line may be packed with another scalar line. */
    boolean canPack;

    /** Whether this line may be joined with another line. */
    boolean canJoin;

    /** Whether this line may participate in a grid. */   
    boolean canGrid;

    static int partsLength(List<String> parts) {
        if (parts.isEmpty()) {
            return 0;
        }

        int length = parts.size() - 1;   // spaces between parts

        for (String part : parts) {
            length += part.length();
        }

        return length;
    }

    void setParts(List<String> parts) {
        this.parts = parts;
        this.partsLength = partsLength(parts);
    }

    Line(int indent) {
        this.indent = indent;
    }

    /**
     * Parse one pretty-printed JSON line.
     *
     * @param s          line text without trailing newline
     * @return parsed line metadata
     */
    static final String EMPTY_OBJECT = "{ }" ;
    static final String EMPTY_OBJECT1 = "{ }," ;
    static final String EMPTY_LIST = "[ ]" ;
    static final String EMPTY_LIST1 = "[ ]," ;

    static Line parse(String s) {

        int start = 0;
        while (start < s.length() && Character.isWhitespace(s.charAt(start))) {
            start++;
        }

        String body = rstrip(s.substring(start));

        Line line = new Line(start) ;

        if (body.endsWith("{")) {
            line.opener = Kind.DICT;
        } else if (body.endsWith("[")) {
            line.opener = Kind.LIST;
        } else if (body.endsWith(EMPTY_OBJECT)) {
            body = body.substring(0, body.length()-EMPTY_OBJECT.length()) + "{}" ;
        } else if (body.endsWith(EMPTY_OBJECT1)) {
            body = body.substring(0, body.length()-EMPTY_OBJECT1.length()) + "{}," ;
        } else if (body.endsWith(EMPTY_LIST)) {
            body = body.substring(0, body.length()-EMPTY_LIST.length()) + "[]" ;
        } else if (body.endsWith(EMPTY_LIST1)){
            body = body.substring(0, body.length()-EMPTY_LIST1.length()) + "[]," ;
        } else {
            line.closer = closingKind(body);
        }

        line.parts.add(body) ;
        line.partsLength = body.length() ;


        boolean isBodyLine = line.opener == Kind.NONE && line.closer == Kind.NONE;

        line.canPack = isBodyLine;
        line.canJoin = isBodyLine;
        line.leafs = isBodyLine ? 1 : 0 ;
        line.items = isBodyLine ? 1 : 0 ;

        return line;
    }

    /**
     * Return this line as output text.
     *
     * @return indentation, line text, and trailing newline
     */
    String raw() {
        return " ".repeat(indent) + String.join(" ", parts) + "\n";
    }

    /**
     * Return physical output width.
     *
     * @return indentation plus text length
     */
    int width() {
        return indent + partsLength;
    }

    boolean canMerge(Line other, int itemLimit, int widthLimit) {
        return (
            this.indent == other.indent
            && this.items + other.items <= itemLimit
            && this.indent + this.partsLength + 1 + other.partsLength <= widthLimit
        );
    }

    /**
     * Merge another line into this line.
     *
     * <p>
     * Used by pack/join phases after width and count checks have passed.
     *
     * @param other line to append
     */
    void mergeLine(Line other) {
        if ( other.parts.isEmpty() ) return ;
        this.parts.addAll(other.parts) ;
        this.partsLength += 1 + other.partsLength;
        items += other.items;
        leafs += other.leafs;

        if (other.childNesting > childNesting) {
            childNesting = other.childNesting;
            canPack = false;
        }
    }

    private static Kind closingKind(String body) {
        return switch (body) {
            case "}", "}," -> Kind.DICT;
            case "]", "]," -> Kind.LIST;
            default -> Kind.NONE;
        };
    }

    private static String rstrip(String s) {
        int end = s.length();
        while (end > 0 && Character.isWhitespace(s.charAt(end - 1))) {
            end--;
        }
        return s.substring(0, end);
    }



    private static final Pattern KEY_RE = Pattern.compile(
            "^\\s*(?:\"[^\"\\\\]*\"|'[^'\\\\]*'|[A-Za-z_$][A-Za-z0-9_$]*)\\s*:"
    );

    List<String> dictSignature() {
        ArrayList<String> signature = new ArrayList<>();

        for (int i = 1; i < parts.size() - 1; i++) {
            Matcher m = KEY_RE.matcher(parts.get(i));
            if (!m.find()) {
                return null;
            }
            signature.add(m.group());
        }

        return signature;
    }

    private static List<String> formatParts(List<String> parts, int[] widths) {
        ArrayList<String> out = new ArrayList<>(parts.size());
        int last = widths.length - 1;

        for (int i = 0; i < parts.size(); i++) {
            String part = parts.get(i);
            if (!part.isEmpty() && "-0123456789".indexOf(part.charAt(0)) >= 0) {
                out.add(" ".repeat(widths[i] - part.length()) + part);
            } else if (i < last) {
                out.add(part + " ".repeat(widths[i] - part.length()));
            } else {
                out.add(part);
            }
        }

        return out;
    }

    void applyGrid(int[] widths) {
        List<String> new_parts = formatParts(parts, widths) ;
        setParts(new_parts);
    }

}

File: dotnet/src/JsonFold/Line.cs
using System.Text.RegularExpressions;

namespace JsonFold;

internal sealed class Line
{
    private static readonly Regex KeyRegex = new(@"^\s*(?:""[^""\\]*""|'[^'\\]*'|[A-Za-z_$][A-Za-z0-9_$]*)\s*:", RegexOptions.Compiled);

    public int Indent { get; }
    public Kind Kind { get; set; } = Kind.None;
    public List<string> Parts { get; private set; } = new();
    public int PartsLength { get; private set; }
    public int Items { get; set; }
    public int Leafs { get; set; }
    public int ChildNesting { get; set; } = -1;
    public Kind Opener { get; set; } = Kind.None;
    public Kind Closer { get; set; } = Kind.None;
    public bool CanPack { get; set; }
    public bool CanJoin { get; set; }
    public bool CanGrid { get; set; }

    internal Line(int indent) => Indent = indent;

    public static Line Parse(string s)
    {
        var start = 0;
        while (start < s.Length && char.IsWhiteSpace(s[start])) start++;
        var body = s[start..].TrimEnd();

        var line = new Line(start);
        line.SetParts(new List<string> { body });

        if (body.EndsWith('{')) line.Opener = Kind.Dict;
        else if (body.EndsWith('[')) line.Opener = Kind.List;

        line.Closer = body switch
        {
            "}" or "}," => Kind.Dict,
            "]" or "]," => Kind.List,
            _ => Kind.None,
        };

        var isBody = line.Opener == Kind.None && line.Closer == Kind.None;
        line.CanPack = isBody;
        line.CanJoin = isBody;
        line.Items = isBody ? 1 : 0;
        line.Leafs = isBody ? 1 : 0;
        return line;
    }

    public string Raw() => new string(' ', Indent) + string.Join(" ", Parts) + "\n";
    public int Width() => Indent + PartsLength;

    public bool CanMerge(Line other, int itemLimit, int widthLimit) =>
        Indent == other.Indent &&
        Items + other.Items <= itemLimit &&
        Indent + PartsLength + 1 + other.PartsLength <= widthLimit;

    public void MergeLine(Line other)
    {
        if (other.Parts.Count == 0) return;
        Parts.AddRange(other.Parts);
        PartsLength += 1 + other.PartsLength;
        Items += other.Items;
        Leafs += other.Leafs;
        if (other.ChildNesting > ChildNesting)
        {
            ChildNesting = other.ChildNesting;
            CanPack = false;
        }
    }

    public void SetParts(List<string> parts)
    {
        Parts = parts;
        PartsLength = CalculatePartsLength(parts);
    }

    public List<string>? DictSignature()
    {
        var signature = new List<string>();
        for (var i = 1; i < Parts.Count - 1; i++)
        {
            var match = KeyRegex.Match(Parts[i]);
            if (!match.Success) return null;
            signature.Add(match.Value);
        }
        return signature;
    }

    public void ApplyGrid(int[] widths) => SetParts(FormatParts(Parts, widths));

    private static int CalculatePartsLength(IReadOnlyCollection<string> parts)
    {
        if (parts.Count == 0) return 0;
        var length = parts.Count - 1;
        foreach (var part in parts) length += part.Length;
        return length;
    }

    private static List<string> FormatParts(IReadOnlyList<string> parts, IReadOnlyList<int> widths)
    {
        var output = new List<string>(parts.Count);
        var last = widths.Count - 1;
        for (var i = 0; i < parts.Count; i++)
        {
            var part = parts[i];
            if (part.Length > 0 && "-0123456789".Contains(part[0]))
                output.Add(part.PadLeft(widths[i]));
            else if (i < last)
                output.Add(part.PadRight(widths[i]));
            else
                output.Add(part);
        }
        return output;
    }
}
